pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/MainDesignerController.java

```diff
@@ -346,9 +316,11 @@ private void updateRecentFilesMenu() {
                 Tooltip.install(item.getContent(), new Tooltip(f.getAbsolutePath()));
                 items.add(item);
             } else {
-                recentFiles.remove(f);
+                filesToClear.add(f);
```

Comment:
An alternative would be, to build up a new list of recent files and replace the list in the end.
Thinking about the problem here - wouldn't it be nice, if we had a rule, that detected if one looped through a list and modifies this list inside the loop?

Reply:
> build up a new list of recent files and replace the list in the end.

That would be nice. I think I wanted to limit assignments to `recentFiles`, but looking back at it, it's not even final. For clarity, we could move that logic into a separate class. State flow in MainController is already very complicated.

> wouldn't it be nice, if we had a rule, that detected if one looped through a list and modifies this list inside the loop?

I totally agree. This pattern is something I do regularly without thinking, making a rule to check for it could be nice. Though, maybe it should be limited to looping on local variables, if field immutability is required. That could be a configurable property of the rule